src/math.F90

```diff
@@ -574,6 +574,111 @@ pure function calc_rn(n,uvw) result(rn)
+                                       ! easier to work with
+    real(8) :: k1, k2, k3, k4          ! Variables for R_m_n calculation
+    real(8) :: sqrt_norm               ! normalization for radial moments
+    integer :: i,p,q                   ! Loop counters
+
+    real(8), parameter :: SQRT_N_1(0:10) = [&
+         sqrt(1.0_8), sqrt(2.0_8), sqrt(3.0_8), sqrt(4.0_8), &
+         sqrt(5.0_8), sqrt(6.0_8), sqrt(7.0_8), sqrt(8.0_8), &
+         sqrt(9.0_8), sqrt(10.0_8), sqrt(11.0_8)]
+    real(8), parameter :: SQRT_2N_2(0:10) = SQRT_N_1*sqrt(2.0_8)
+
+    ! n == radial degree
+    ! m == azimuthal frequency
+
+    ! ==========================================================================
+    ! Determine vector of sin(n*phi) and cos(n*phi). This takes advantage of the
+    ! following recurrence relations so that only a single sin/cos have to be
+    ! evaluated (http://mathworld.wolfram.com/Multiple-AngleFormulas.html)
+    !
+    ! sin(nx) = 2 cos(x) sin((n-1)x) - sin((n-2)x)
+    ! cos(nx) = 2 cos(x) cos((n-1)x) - cos((n-2)x)
+
+    sin_phi = sin(phi)
+    cos_phi = cos(phi)
+
+    sin_phi_vec(1) = 1.0_8
+    cos_phi_vec(1) = 1.0_8
+
+    sin_phi_vec(2) = 2.0_8 * cos_phi
+    cos_phi_vec(2) = cos_phi
+
+    do i = 3, n+1
+      sin_phi_vec(i) = 2.0_8 * cos_phi * sin_phi_vec(i-1) - sin_phi_vec(i-2)
+      cos_phi_vec(i) = 2.0_8 * cos_phi * cos_phi_vec(i-1) - cos_phi_vec(i-2)
+    end do
+
+    do i = 1, n+1
+      sin_phi_vec(i) = sin_phi_vec(i) * sin_phi
+    end do
+
+    ! ==========================================================================
+    ! Calculate R_pq(rho)
+
+    ! Fill the main diagonal first (Eq. 3.9 in Chong)
+    do p = 0, n
+      zn_mat(p+1, p+1) = rho**p
+    end do
+
+    ! Fill in the second diagonal (Eq. 3.10 in Chong)
+    do q = 0, n-2
+      zn_mat(q+2+1, q+1) = (q+2) * zn_mat(q+2+1, q+2+1) - (q+1) * zn_mat(q+1, q+1)
+    end do
+
+    ! Fill in the rest of the values using the original results (Eq. 3.8 in Chong)
+    do p = 4, n
+      k2 = 2 * p * (p - 1) * (p - 2)
+      do q = p-4, 0, -2
+        k1 = (p + q) * (p - q) * (p - 2) / 2
+        k3 = -q**2*(p - 1) - p * (p - 1) * (p - 2)
+        k4 = -p * (p + q - 2) * (p - q - 2) / 2
+        zn_mat(p+1, q+1) = ((k2 * rho**2 + k3) * zn_mat(p-2+1, q+1) + k4 * zn_mat(p-4+1, q+1)) / k1
+      end do
+    end do
+
+    ! Roll into a single vector for easier computation later
+    ! The vector is ordered (0,0), (1,-1), (1,1), (2,-2), (2,0),
+    ! (2, 2), ....   in (n,m) indices
+    ! Note that the cos and sin vectors are offset by one
+    ! sin_phi_vec = [sin(x), sin(2x), sin(3x) ...]
+    ! cos_phi_vec = [1.0, cos(x), cos(2x)... ]
+    i = 1
+    do p = 0, n
+      do q = -p, p, 2
+        if (q < 0) then
+          zn(i) = zn_mat(p+1, abs(q)+1) * sin_phi_vec(abs(q)) * SQRT_2N_2(p)
+        else if (q == 0) then
+          zn(i) = zn_mat(p+1, q+1) * SQRT_N_1(p)
+        else
+          zn(i) = zn_mat(p+1, q+1) * cos_phi_vec(abs(q)+1) * SQRT_2N_2(p)
```

Comment:
I'm a little confused here. Correct me if I'm wrong, but we're calculating and `Z_1^1 = rho * cos(phi) * sqrt(2)`? 

Integrating `Z^2 * rho * drho * dphi` over `0 < rho < 1` and `0 < phi < 2*pi` gives `pi / 2` for the `Z_1^1` case. I'm confused as to what normalization we're using here. Wikipedia shows a couple - the 5th equation under the 'Orthogonality' heading would suggest that our orthogonal integral for `Z_1^1` should evaluate to `pi/4`, while the first equation under heading `Zernike polynomials` followed by the text:

> The first few Zernike modes, with OSA/ANSI and Noll single-indices...

suggests the integral should equal `pi`. But unless I made a math error our integral of `pi/2` doesn't match either of these?

Reply:
`Z_1^1 = rho * cos(phi) * 2` not sqrt(2), so when it gets squared in the integrand you end up with another factor of 2, hence an integral of pi.

Reply:
But your point stands that the normalization is not clear. We're using the modified Zernike polynomials where it has already been multipled by the sqrt(2n+ 2) or sqrt(n+1) term so that they always integrate to pi. I'll adjust the documentation accordingly.

Reply:
Ah, got it, thanks! Then Matt's original implementation used orthonormal Zernike polynomials, since each was divided by `SQRT_PI`.

Reply:
Wait, but isn't `SQRT_2N_2(1) = SQRT_N_1(1) * sqrt(2.0_8)`, so the scalar in `Z_1^1` would be `sqrt(2)`, since `p = 1` for this polynomial?

Reply:
it's sqrt(4) because those arrays are explicitly indexed from zero

Reply:
My Fortran is a little rusty, thanks!